src/main/java/bio/terra/service/filedata/google/firestore/FireStoreFileDao.java

```diff
@@ -75,17 +75,31 @@ void createFileMetadata(Firestore firestore, String datasetId, FireStoreFile new
    * Upserts file metadata objects into Firestore (e.g. this is the metadata that contains size,
    * checksum, cloud location etc.) of a file, as opposed to the path information for the file
    */
-  void createFileMetadata(Firestore firestore, String datasetId, List<FireStoreFile> newFiles)
+  void upsertFileMetadata(Firestore firestore, String datasetId, List<FireStoreFile> newFiles)
       throws InterruptedException {
     String collectionId = makeCollectionId(datasetId);
     fireStoreUtils.batchOperation(
         newFiles,
         newFile ->
-            firestore.runTransaction(
-                xn -> {
-                  xn.set(getFileDocRef(firestore, collectionId, newFile.getFileId()), newFile);
-                  return null;
-                }));
+            firestore.runTransaction(upsertSingleFileFunction(firestore, collectionId, newFile)));
+  }
+
+  private <T> Transaction.Function<T> upsertSingleFileFunction(
+      Firestore firestore, String collectionId, FireStoreFile newFile) {
+    return xn -> {
+      DocumentReference fileDocRef = getFileDocRef(firestore, collectionId, newFile.getFileId());
+      DocumentSnapshot documentSnapshot = xn.get(fileDocRef).get();
+      if (documentSnapshot.exists()) {
+        FireStoreFile existingFile = documentSnapshot.toObject(FireStoreFile.class);
+        if (!Objects.equals(existingFile.getLoadTag(), newFile.getLoadTag())) {
```

Comment:
We won't run into problems with this check, right? Since we will be uploading the existing file with the only difference being a specified md5 (so any load tag on the existing will be present on the new)?

Reply:
Yeah exactly. I added this for future-proofing since this methods lives in the Dao so will be longer lived.  In our current usage it's fine but I could see it failing down the road